pkg/executor/statefulset.go

```diff
@@ -149,26 +149,52 @@ func UpdateStatefulSetNonReplicas(
 	// need/expect to be under our control, other than the replicas count,
 	// correct them here.
 
-	// For now only checking the owner reference.
-	if shared.OwnerReferencesPresent(cr, statefulSet.OwnerReferences) {
+	patchedRes := *statefulSet
+	needPatch := false
+
+	// Check the owner reference.
+	if !shared.OwnerReferencesPresent(cr, statefulSet.OwnerReferences) {
+		shared.LogInfof(
+			reqLogger,
+			cr,
+			shared.EventReasonNoEvent,
+			"repairing owner ref on statefulset{%s}",
+			statefulSet.Name,
+		)
+
+		// So, what to do. Do we add our owner ref to the existing ones? What if
+		// something else is claiming to be controller? Probably some stale ref
+		// left by a bad backup/restore process? We're just going to nuke any
+		// existing owner refs.
+		patchedRes.OwnerReferences = shared.OwnerReferences(cr)
+		needPatch = true
+	}
+
+	// EZML-862
+	// We should also compare the current statefulset container image against
+	// the image defined for the corresponding role in the KDApp spec. If images
+	// are not the same, KDApp spec image has a higher priority, so we shoud
+	// upgrade the statefulset one.
+	// Make sure, that according this logic, there is no sence to edit a statefulset
+	// directly, as it will be reconciled back to the KDApp spec state.
+	appRoleImage, err := catalog.ImageForRole(cr, role.Name)
+	if err != nil {
+		return err
+	}
+
+	ssImage := catalog.ImageForStatefulSet(statefulSet)
+
+	if strings.Compare(appRoleImage, ssImage) != 0 {
+		patchedRes.Spec.Template.Spec.Containers[0].Image = appRoleImage
+		needPatch = true
+	}
+
+	if !needPatch {
 		return nil
 	}
-	shared.LogInfof(
-		reqLogger,
-		cr,
-		shared.EventReasonNoEvent,
-		"repairing owner ref on statefulset{%s}",
-		statefulSet.Name,
-	)
-	// So, what to do. Do we add our owner ref to the existing ones? What if
-	// something else is claiming to be controller? Probably some stale ref
-	// left by a bad backup/restore process? We're just going to nuke any
-	// existing owner refs.
-	patchedRes := *statefulSet
-	patchedRes.OwnerReferences = shared.OwnerReferences(cr)
-	patchErr := shared.Patch(
+
+	patchErr := shared.Update(
```

Comment:
I'm really puzzled why Patch didn't work for you... we'd need to explore this a bit more.

A problem with Update is that it will fail if the resource has been modified since it has been read (either by some other code in KD or by the K8s statefulset controller or by whoever). This is not rare, and it means that the whole handler will fail and the next handler will have to try again. This is annoying at best, and at worst it's possible with some kinds of updates to get hardlocked and never succeed because (for this or that reason) one part of KD always updates the resource before some other part tries.

Anyway... patch is nice! Let's see if we can get it to work.

Reply:
OK so yeah, I think it's a shallow-copy problem. On the line above we just do "patchedRes := *statefulSet" which only copies over the top-level properties of the statefulset in order to make the new object. This was fine when we were just updating the owner ref, but now that we're digging down into the slice of containers it's not fine. When we modify the containers slice, the modification is seen by both the old and new object.

So before doing that modification, we need to deep-copy the container slice and assign that copy to the patchedRes. One example of doing that sort of thing is for the annotations slice modified by SetBackupAnnotation in pkg/executor/cluster.go.

Reply:
Got it, I changed the container-related logic and it looks like that `Patch` works